rustdoc-search: shard the search result descriptions

notriddle opened the pull request:
## Preview

This makes no visual changes to rustdoc search. It's a pure perf improvement.

<details><summary>old</summary>

Preview: <http://notriddle.com/rustdoc-html-demo-10/doc/std/index.html?search=vec>

WebPageTest Comparison with before branch on a sort of worst case (searching `vec`, winds up downloading most of the shards anyway): <https://www.webpagetest.org/video/compare.php?tests=240317_AiDc61_2EM,240317_AiDcM0_2EN>

Waterfall diagram:
![image](https://github.com/rust-lang/rust/assets/1593513/39548f0c-7ad6-411b-abf8-f6668ff4da18)

</details>

Preview: <http://notriddle.com/rustdoc-html-demo-10/doc2/std/index.html?search=vec>

WebPageTest Comparison with before branch on a sort of worst case (searching `vec`, winds up downloading most of the shards anyway): <https://www.webpagetest.org/video/compare.php?tests=240322_BiDcCH_13R,240322_AiDcJY_104>

![image](https://github.com/rust-lang/rust/assets/1593513/4be1f9ff-c3ff-4b96-8f5b-b264df2e662d)

## Description

r? @GuillaumeGomez 

The descriptions are, on almost all crates[^1], the majority of the size of the search index, even though they aren't really used for searching. This makes it relatively easy to separate them into their own files.

Additionally, this PR pulls out information about whether there's a description into a bitmap. This allows us to sort, truncate, *then* download.

This PR also bumps us to ES8. Out of the browsers we support, all of them support async functions according to caniuse.

https://caniuse.com/async-functions

[^1]:
    <https://microsoft.github.io/windows-docs-rs/>, a crate with
    44MiB of pure names and no descriptions for them, is an outlier
    and should not be counted. But this PR should improve it, by replacing a long line of empty strings with a compressed bitmap with a single Run section. Just not very much.

## Detailed sizes

```console
$ cat test.sh 
set -ex
cp ../search-index*.js search-index.js
awk 'FNR==NR {a++;next} FNR<a-3' search-index.js{,} | awk 'NR>1 {gsub(/\],\\$/,""); gsub(/^\["[^"]+",/,""); print} {next}' | sed -E "s:\\\\':':g" > search-index.json
jq -c '.t' search-index.json > t.json
jq -c '.n' search-index.json > n.json
jq -c '.q' search-index.json > q.json
jq -c '.D' search-index.json > D.json
jq -c '.e' search-index.json > e.json
jq -c '.i' search-index.json > i.json
jq -c '.f' search-index.json > f.json
jq -c '.c' search-index.json > c.json
jq -c '.p' search-index.json > p.json
jq -c '.a' search-index.json > a.json
du -hs t.json n.json q.json D.json e.json i.json f.json c.json p.json a.json
$ bash test.sh 
+ cp ../search-index1.78.0.js search-index.js
+ awk 'FNR==NR {a++;next} FNR<a-3' search-index.js search-index.js
+ awk 'NR>1 {gsub(/\],\\$/,""); gsub(/^\["[^"]+",/,""); print} {next}'
+ sed -E 's:\\'\'':'\'':g'
+ jq -c .t search-index.json
+ jq -c .n search-index.json
+ jq -c .q search-index.json
+ jq -c .D search-index.json
+ jq -c .e search-index.json
+ jq -c .i search-index.json
+ jq -c .f search-index.json
+ jq -c .c search-index.json
+ jq -c .p search-index.json
+ jq -c .a search-index.json
+ du -hs t.json n.json q.json D.json e.json i.json f.json c.json p.json a.json
64K     t.json
800K    n.json
8.0K    q.json
4.0K    D.json
16K     e.json
192K    i.json
544K    f.json
4.0K    c.json
36K     p.json
20K     a.json
```

These are, roughly, the size of each section in the standard library (this tool actually excludes libtest, for parsing-json-with-awk reasons, but libtest is tiny so it's probably not important).

t = item type, like "struct", "free fn", or "type alias". Since one byte is used for every item, this implies that there are approximately 64 thousand items in the standard library.

n = name, and that's now the largest section of the search index with the descriptions removed from it

q = parent *module* path, stored parallel to the items within

D = the size of each description shard, stored as vlq hex numbers

e = empty description bit flags, stored as a roaring bitmap

i = parent *type* index as a link into `p`, stored as decimal json numbers; used only for associated types; might want to switch to vlq hex, since that's shorter, but that would be a separate pr

f = function signature, stored as lists of lists that index into `p`

c = deprecation flag, stored as a roaring bitmap

p = parent *type*, stored separately and linked into from `i` and `f`

a = alias, as [[key, value]] pairs

## Search performance

http://notriddle.com/rustdoc-html-demo-11/perf-shard/index.html

For example, in stm32f4:

<table><thead><tr><th>before<th>after</tr></thead>
<tbody><tr><td>

```
Testing T -> U ... in_args = 0, returned = 0, others = 200
wall time = 617

Testing T, U ... in_args = 0, returned = 0, others = 200
wall time = 198

Testing T -> T ... in_args = 0, returned = 0, others = 200
wall time = 282

Testing crc32 ... in_args = 0, returned = 0, others = 0
wall time = 426

Testing spi::pac ... in_args = 0, returned = 0, others = 0
wall time = 673
```

</td><td>

```
Testing T -> U ... in_args = 0, returned = 0, others = 200
wall time = 716

Testing T, U ... in_args = 0, returned = 0, others = 200
wall time = 207

Testing T -> T ... in_args = 0, returned = 0, others = 200
wall time = 289

Testing crc32 ... in_args = 0, returned = 0, others = 0
wall time = 418

Testing spi::pac ... in_args = 0, returned = 0, others = 0
wall time = 687
```

</td></tr><tr><td>

```
user: 005.345 s
sys:  002.955 s
wall: 006.899 s
child_RSS_high:     583664 KiB
group_mem_high:     557876 KiB
```

</td><td>

```
user: 004.652 s
sys:  000.565 s
wall: 003.865 s
child_RSS_high:     538696 KiB
group_mem_high:     511724 KiB
```

</td></tr>

</table>

This perf tester is janky and unscientific enough that the apparent differences might just be noise. If it's not an order of magnitude, it's probably not real.

## Future possibilities

* Currently, results are not shown until the descriptions are downloaded. Theoretically, the description-less results could be shown. But actually doing that, and making sure it works properly, would require extra work (we have to be careful to avoid layout jumps).
* More than just descriptions can be sharded this way. But we have to be careful to make sure the size wins are worth the round trips. Ideally, data that’s needed only for display should be sharded while data needed for search isn’t.
* [Full text search](https://internals.rust-lang.org/t/full-text-search-for-rustdoc-and-doc-rs/20427) also needs this kind of infrastructure. A good implementation might store a compressed bloom filter in the search index, then download the full keyword in shards. But, we have to be careful not just of the amount readers have to download, but also of the amount that [publishers](https://gist.github.com/notriddle/c289e77f3ed469d1c0238d1d135d49e1) have to store.

notriddle committed:
rustdoc: clean up formatting

notriddle committed:
rustdoc-search: shard the search result descriptions

The descriptions are, on almost all crates[^1], the majority
of the size of the search index, even though they aren't really
used for searching. This makes it relatively easy to separate
them into their own files.

This commit also bumps us to ES8. Out of the browsers we support,
all of them support async functions according to caniuse.

https://caniuse.com/async-functions

[^1]:
    <https://microsoft.github.io/windows-docs-rs/>, a crate with
    44MiB of pure names and no descriptions for them, is an outlier
    and should not be counted.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8312518476/job/22747474884) [(plain)](https://github.com/rust-lang/rust/commit/b731ed662120b093b5850141f084202f2cd0acb6/checks/22747474884/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=notriddle
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_581a862e-29ff-4c56-9118-74ea9587b157
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=notriddle/search-desc
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_581a862e-29ff-4c56-9118-74ea9587b157
GITHUB_REF=refs/pull/122614/merge
GITHUB_REF_NAME=122614/merge
GITHUB_REF_PROTECTED=false
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=7376c96cf5dac46f5b9b1efc32e566de2d215929
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_581a862e-29ff-4c56-9118-74ea9587b157
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_581a862e-29ff-4c56-9118-74ea9587b157
GITHUB_TRIGGERING_ACTOR=notriddle
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/122614/merge
GITHUB_WORKFLOW_SHA=7376c96cf5dac46f5b9b1efc32e566de2d215929
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_20_X64=/opt/hostedtoolcache/go/1.20.14/x64
---
          · error: SyntaxError: Unexpected token (333:24)
          · see the printed err.stack below for context
          ----

          SyntaxError: Unexpected token (333:24)
    at Parser.pp$4.raise (/node-v16.9.0-linux-x64/lib/node_modules/es-check/node_modules/acorn/dist/acorn.js:3573:15)
    at Parser.pp$9.unexpected (/node-v16.9.0-linux-x64/lib/node_modules/es-check/node_modules/acorn/dist/acorn.js:772:10)
    at Parser.pp$9.expect (/node-v16.9.0-linux-x64/lib/node_modules/es-check/node_modules/acorn/dist/acorn.js:766:28)
    at Parser.pp$5.parseObj (/node-v16.9.0-linux-x64/lib/node_modules/es-check/node_modules/acorn/dist/acorn.js:3182:14)
    at Parser.pp$5.parseExprAtom (/node-v16.9.0-linux-x64/lib/node_modules/es-check/node_modules/acorn/dist/acorn.js:2907:19)
    at Parser.pp$5.parseExprSubscripts (/node-v16.9.0-linux-x64/lib/node_modules/es-check/node_modules/acorn/dist/acorn.js:2715:21)
    at Parser.pp$5.parseMaybeUnary (/node-v16.9.0-linux-x64/lib/node_modules/es-check/node_modules/acorn/dist/acorn.js:2681:19)
    at Parser.pp$5.parseExprOps (/node-v16.9.0-linux-x64/lib/node_modules/es-check/node_modules/acorn/dist/acorn.js:2608:21)
    at Parser.pp$5.parseMaybeConditional (/node-v16.9.0-linux-x64/lib/node_modules/es-check/node_modules/acorn/dist/acorn.js:2591:21)
    at Parser.pp$5.parseMaybeAssign (/node-v16.9.0-linux-x64/lib/node_modules/es-check/node_modules/acorn/dist/acorn.js:2558:21)
info: 
          ES-Check Error:
          ----
          · erroring file: ../src/librustdoc/html/static/js/search.js
          · erroring file: ../src/librustdoc/html/static/js/search.js
          · error: SyntaxError: Unexpected token (210:16)
          · see the printed err.stack below for context
          ----

          SyntaxError: Unexpected token (210:16)
    at Parser.pp$4.raise (/node-v16.9.0-linux-x64/lib/node_modules/es-check/node_modules/acorn/dist/acorn.js:3573:15)
    at Parser.pp$9.unexpected (/node-v16.9.0-linux-x64/lib/node_modules/es-check/node_modules/acorn/dist/acorn.js:772:10)
    at Parser.pp$5.parseExprAtomDefault (/node-v16.9.0-linux-x64/lib/node_modules/es-check/node_modules/acorn/dist/acorn.js:2936:10)
    at Parser.pp$5.parseExprAtom (/node-v16.9.0-linux-x64/lib/node_modules/es-check/node_modules/acorn/dist/acorn.js:2931:19)
    at Parser.pp$5.parseExprSubscripts (/node-v16.9.0-linux-x64/lib/node_modules/es-check/node_modules/acorn/dist/acorn.js:2715:21)
    at Parser.pp$5.parseMaybeUnary (/node-v16.9.0-linux-x64/lib/node_modules/es-check/node_modules/acorn/dist/acorn.js:2681:19)
    at Parser.pp$5.parseExprOps (/node-v16.9.0-linux-x64/lib/node_modules/es-check/node_modules/acorn/dist/acorn.js:2608:21)
    at Parser.pp$5.parseMaybeConditional (/node-v16.9.0-linux-x64/lib/node_modules/es-check/node_modules/acorn/dist/acorn.js:2591:21)
    at Parser.pp$5.parseMaybeAssign (/node-v16.9.0-linux-x64/lib/node_modules/es-check/node_modules/acorn/dist/acorn.js:2558:21)
    at Parser.pp$5.parseExprList (/node-v16.9.0-linux-x64/lib/node_modules/es-check/node_modules/acorn/dist/acorn.js:3444:20)
info: 
          ES-Check Error:
          ----
          · erroring file: ../src/librustdoc/html/static/js/storage.js
          · erroring file: ../src/librustdoc/html/static/js/storage.js
          · error: SyntaxError: Unexpected token (215:8)
          · see the printed err.stack below for context
          ----

          SyntaxError: Unexpected token (215:8)
    at Parser.pp$4.raise (/node-v16.9.0-linux-x64/lib/node_modules/es-check/node_modules/acorn/dist/acorn.js:3573:15)
    at Parser.pp$9.unexpected (/node-v16.9.0-linux-x64/lib/node_modules/es-check/node_modules/acorn/dist/acorn.js:772:10)
    at Parser.pp$5.parseExprAtomDefault (/node-v16.9.0-linux-x64/lib/node_modules/es-check/node_modules/acorn/dist/acorn.js:2936:10)
    at Parser.pp$5.parseExprAtom (/node-v16.9.0-linux-x64/lib/node_modules/es-check/node_modules/acorn/dist/acorn.js:2931:19)
    at Parser.pp$5.parseExprSubscripts (/node-v16.9.0-linux-x64/lib/node_modules/es-check/node_modules/acorn/dist/acorn.js:2715:21)
    at Parser.pp$5.parseMaybeUnary (/node-v16.9.0-linux-x64/lib/node_modules/es-check/node_modules/acorn/dist/acorn.js:2681:19)
    at Parser.pp$5.parseExprOps (/node-v16.9.0-linux-x64/lib/node_modules/es-check/node_modules/acorn/dist/acorn.js:2608:21)
    at Parser.pp$5.parseMaybeConditional (/node-v16.9.0-linux-x64/lib/node_modules/es-check/node_modules/acorn/dist/acorn.js:2591:21)
    at Parser.pp$5.parseMaybeAssign (/node-v16.9.0-linux-x64/lib/node_modules/es-check/node_modules/acorn/dist/acorn.js:2558:21)
    at Parser.pp$5.parseExprList (/node-v16.9.0-linux-x64/lib/node_modules/es-check/node_modules/acorn/dist/acorn.js:3444:20)
  local time: Sun Mar 17 03:20:44 UTC 2024
  network time: Sun, 17 Mar 2024 03:20:44 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8313116433/job/22748768475) [(plain)](https://github.com/rust-lang/rust/commit/5b44bfda7fc62b2874400e613672aefe5b49aaaa/checks/22748768475/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=notriddle
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_a0feff60-055b-4d62-b7d2-dbb9bd313c09
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=notriddle/search-desc
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_a0feff60-055b-4d62-b7d2-dbb9bd313c09
GITHUB_REF=refs/pull/122614/merge
GITHUB_REF_NAME=122614/merge
GITHUB_REF_PROTECTED=false
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=f96b6c80e81546cee7201b14744c7adeec19e09b
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_a0feff60-055b-4d62-b7d2-dbb9bd313c09
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_a0feff60-055b-4d62-b7d2-dbb9bd313c09
GITHUB_TRIGGERING_ACTOR=notriddle
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/122614/merge
GITHUB_WORKFLOW_SHA=f96b6c80e81546cee7201b14744c7adeec19e09b
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_20_X64=/opt/hostedtoolcache/go/1.20.14/x64
---
#15 3.446 Building wheels for collected packages: reuse
#15 3.447   Building wheel for reuse (pyproject.toml): started
#15 3.774   Building wheel for reuse (pyproject.toml): finished with status 'done'
#15 3.775   Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=181117 sha256=f5f58750481f69515c2c0d1d503daf565e2565c370d07fc6aeb95fe3498b4269
#15 3.775   Stored in directory: /tmp/pip-ephem-wheel-cache-5zkda594/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
#15 3.778 Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
#15 3.802   Attempting uninstall: setuptools
#15 3.803     Found existing installation: setuptools 59.6.0
#15 3.804     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
---
info: ES-Check: there were no ES version matching errors!  🎉
+ eslint -c ../src/librustdoc/html/static/.eslintrc.js ../src/librustdoc/html/static/js/externs.js ../src/librustdoc/html/static/js/main.js ../src/librustdoc/html/static/js/scrape-examples.js ../src/librustdoc/html/static/js/search.js ../src/librustdoc/html/static/js/settings.js ../src/librustdoc/html/static/js/src-script.js ../src/librustdoc/html/static/js/storage.js

/checkout/src/librustdoc/html/static/js/main.js
  349:34  error  Unexpected space before function parentheses  space-before-function-paren

✖ 1 problem (1 error, 0 warnings)
  1 error and 0 warnings potentially fixable with the `--fix` option.
  local time: Sun Mar 17 05:28:54 UTC 2024
  network time: Sun, 17 Mar 2024 05:28:54 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

notriddle committed:
Fix style errors

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8313374892/job/22749284209) [(plain)](https://github.com/rust-lang/rust/commit/4835b8e218d212568bd40cc0781ab2da6bb1623a/checks/22749284209/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=notriddle
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_38c935aa-1435-4439-8b00-6ea291ff26f3
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=notriddle/search-desc
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_38c935aa-1435-4439-8b00-6ea291ff26f3
GITHUB_REF=refs/pull/122614/merge
GITHUB_REF_NAME=122614/merge
GITHUB_REF_PROTECTED=false
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=5cc533f5adc333a07a797d2b71bdcf8f2bef7911
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_38c935aa-1435-4439-8b00-6ea291ff26f3
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_38c935aa-1435-4439-8b00-6ea291ff26f3
GITHUB_TRIGGERING_ACTOR=notriddle
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/122614/merge
GITHUB_WORKFLOW_SHA=5cc533f5adc333a07a797d2b71bdcf8f2bef7911
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_20_X64=/opt/hostedtoolcache/go/1.20.14/x64
---
#15 3.382 Building wheels for collected packages: reuse
#15 3.383   Building wheel for reuse (pyproject.toml): started
#15 3.703   Building wheel for reuse (pyproject.toml): finished with status 'done'
#15 3.704   Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=181117 sha256=f5f58750481f69515c2c0d1d503daf565e2565c370d07fc6aeb95fe3498b4269
#15 3.704   Stored in directory: /tmp/pip-ephem-wheel-cache-ldvf9bel/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
#15 3.707 Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
#15 3.729   Attempting uninstall: setuptools
#15 3.729     Found existing installation: setuptools 59.6.0
#15 3.730     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
---
+ eslint -c ../src/librustdoc/html/static/.eslintrc.js ../src/librustdoc/html/static/js/externs.js ../src/librustdoc/html/static/js/main.js ../src/librustdoc/html/static/js/scrape-examples.js ../src/librustdoc/html/static/js/search.js ../src/librustdoc/html/static/js/settings.js ../src/librustdoc/html/static/js/src-script.js ../src/librustdoc/html/static/js/storage.js
+ eslint -c ../src/tools/rustdoc-js/.eslintrc.js ../src/tools/rustdoc-js/tester.js

/checkout/src/tools/rustdoc-js/tester.js
  429:34  error  Unexpected space before function parentheses  space-before-function-paren

✖ 1 problem (1 error, 0 warnings)
  1 error and 0 warnings potentially fixable with the `--fix` option.
  local time: Sun Mar 17 06:22:08 UTC 2024
  network time: Sun, 17 Mar 2024 06:22:08 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

GuillaumeGomez commented on src/librustdoc/html/static/js/search.js:
```diff
@@ -2478,9 +2486,9 @@ function initSearch(rawSearchIndex) {
-            sortResults(results_in_args, true, currentCrate),
-            sortResults(results_returned, true, currentCrate),
-            sortResults(results_others, false, currentCrate),
+            await sortResults(results_in_args, true, currentCrate),
```
Instead of doing it one after the other, can't you await on all of them at once?

GuillaumeGomez commented:
Do you have the time difference between storing all docs in one file versus max 1MB per shard?

notriddle committed:
Use promise.all to load sorted results in parallel

GuillaumeGomez commented on src/librustdoc/html/static/js/search.js:
```diff
@@ -2699,9 +2712,9 @@ ${item.displayPath}<span class="${type}">${name}</span>\
-        const ret_others = addTab(results.others, results.query, true);
-        const ret_in_args = addTab(results.in_args, results.query, false);
-        const ret_returned = addTab(results.returned, results.query, false);
+        const ret_others = await addTab(results.others, results.query, true);
```
Can you await them all at once please?

GuillaumeGomez commented on src/tools/rustdoc-js/tester.js:
```diff
@@ -494,21 +533,34 @@ function main(argv) {
+    process.exit(1);
+});
+
+/*process.on("uncaughtException", (err) => {
```
Forgotten commented code?

GuillaumeGomez commented:
Wouldn't it be easier to have an optional index to the position of the item's documentation? Then based on that index, you know which file to download. (Or maybe it's exactly what you said and I completely misunderstood you.)

notriddle committed:
rustdoc-search: compressed bitmap to sort, then load desc

This adds a bit more data than "pure sharding" by
including information about which items have no description
at all. This way, it can sort the results, then truncate,
then finally download the description.

With the "e" bitmap: 2380KiB

Without the "e" bitmap: 2364KiB

GuillaumeGomez commented on src/librustdoc/html/render/search_index.rs:
```diff
@@ -454,42 +471,323 @@ pub(crate) fn build_index<'tcx>(
+        desc.iter().map(|(len, _)| *len).sum::<usize>() + emptydesc.len()
+    );
+
+    // The index, which is actually used to search, is JSON
```
```suggestion
    // The index, which is actually used to search, is JSON.
```

(nitpicking ^^')

GuillaumeGomez commented on src/librustdoc/html/render/search_index.rs:
```diff
@@ -17,12 +18,25 @@ use crate::html::format::join_with_double_colon;
+///
+/// The `index` is a JSON-encoded list of names and other information.
+///
+/// The desc has newlined descriptions, split up by size into 128KiB shards.
```
Would be nice if you added some explanations why `128KiB` was picked and not another.

GuillaumeGomez commented on src/librustdoc/html/render/search_index.rs:
```diff
@@ -17,12 +18,25 @@ use crate::html::format::join_with_double_colon;
+/// The desc has newlined descriptions, split up by size into 128KiB shards.
+/// For example, `(4, "foo\nbar\nbaz\nquux")`.
+pub(crate) struct SerializedSearchIndex {
+    pub(crate) index: String,
```
Is this index actually useful? If you know the index of the item you're looking for (since it's in the `search-index.js` file), you can access it directly into the list, right? Did I misunderstood something?

GuillaumeGomez replied:
I see, I misunderstood what it was.

GuillaumeGomez commented on src/librustdoc/html/render/search_index.rs:
```diff
@@ -454,42 +471,323 @@ pub(crate) fn build_index<'tcx>(
+            crate_data.serialize_field("D", &self.descindex)?;
             crate_data.serialize_field("p", &paths)?;
             crate_data.serialize_field("b", &self.associated_item_disambiguators)?;
+            let mut buf = Vec::new();
+            let mut strbuf = String::new();
+            write_bitmap_to_bytes(&deprecated, &mut buf).unwrap();
+            BASE64_STANDARD.encode_string(&buf, &mut strbuf);
+            crate_data.serialize_field("c", &strbuf)?;
+            strbuf.clear();
+            buf.clear();
+            write_bitmap_to_bytes(&self.emptydesc, &mut buf).unwrap();
+            BASE64_STANDARD.encode_string(&buf, &mut strbuf);
+            crate_data.serialize_field("e", &strbuf)?;
```
Can you put this part into a function please? This code is a bit indigest as is. Having a function name will be helpful to understand better what's going on.

GuillaumeGomez commented on src/librustdoc/html/render/search_index.rs:
```diff
@@ -454,42 +471,323 @@ pub(crate) fn build_index<'tcx>(
-    // Collect the index into a string
-    format!(
+    let (emptydesc, desc) = {
+        let mut emptydesc = Vec::new();
```
`empty_desc`?

GuillaumeGomez commented on src/librustdoc/html/render/search_index.rs:
```diff
@@ -454,42 +471,323 @@ pub(crate) fn build_index<'tcx>(
+        let mut result = Vec::new();
+        let mut set = String::new();
+        let mut len: usize = 0;
+        let mut itemindex: u32 = 0;
```
`item_index`?

GuillaumeGomez commented on src/librustdoc/html/render/search_index.rs:
```diff
@@ -454,42 +471,323 @@ pub(crate) fn build_index<'tcx>(
+        (emptydesc, result)
+    };
+
+    let descindex = {
```
`desc_index`?

GuillaumeGomez commented on src/librustdoc/html/render/search_index.rs:
```diff
@@ -454,42 +471,323 @@ pub(crate) fn build_index<'tcx>(
+    SerializedSearchIndex { index, desc }
+}
+
+pub(crate) fn write_vlqhex_to_string(n: i32, string: &mut String) {
```
Can you move this function and `write_bitmap_to_bytes` into another file please? With a nice doc comment on the module giving extra details about this algorithm.

GuillaumeGomez commented on src/librustdoc/html/render/search_index.rs:
```diff
@@ -454,42 +471,323 @@ pub(crate) fn build_index<'tcx>(
+    // https://arxiv.org/pdf/1603.06549.pdf
+    let mut keys = Vec::<u16>::new();
+    let mut containers = Vec::<Container>::new();
+    enum Container {
```
That will also allow you to move this enum out of the function (makes the code a bit more clear).

GuillaumeGomez commented on src/librustdoc/html/static/js/main.js:
```diff
@@ -329,6 +329,26 @@ function preLoadCss(cssUrl) {
+        descShards: new Map(),
+        loadDesc: async function({descShard, descIndex}) {
+            if (descShard.promise === null) {
+                descShard.promise = new Promise((resolve, reject) => {
```
I don't understand how the promise is working here. There is nothing to await, or did I miss something? We could update `loadScript` to make it `async` too.

GuillaumeGomez replied:
Thanks for adding the comment. :+1:

GuillaumeGomez commented on src/librustdoc/html/static/js/search.js:
```diff
@@ -1446,7 +1454,16 @@ function initSearch(rawSearchIndex) {
+                    searchState.loadDesc(result);
+            }
+            for (const result of transformed) {
+                result.desc = await result.desc;
```
Please await in parallel. :)

notriddle committed:
rustdoc-search: address nits

GuillaumeGomez commented on src/librustdoc/html/render/search_index/encode.rs:
```diff
@@ -0,0 +1,245 @@
+                    }
+                    prev = Some(value);
+                }
+                if 2 + 4 * r < 2 * array.len() + 2 {
```
Same could be done here.

GuillaumeGomez reviewed (commented):
Looks good to me. r=me with or without the suggestions I made.

notriddle committed:
Clean up src/librustdoc/html/render/search_index/encode.rs

Co-authored-by: Guillaume Gomez <guillaume1.gomez@gmail.com>

bors commented:
:pushpin: Commit a272007a209a4327ad8122af1ad9f9620f04724e has been approved by `GuillaumeGomez`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=GuillaumeGomez a272007a209a4327ad8122af1ad9f9620f04724e -->
<!-- homu: {"type":"Approved","sha":"a272007a209a4327ad8122af1ad9f9620f04724e","approver":"GuillaumeGomez","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
